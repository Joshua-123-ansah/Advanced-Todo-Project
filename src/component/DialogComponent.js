import React, { useEffect, useState } from "react";
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  FormControl,
  Grid,
  InputLabel,
  makeStyles,
  MenuItem,
  Select,
  TextField,
} from "@material-ui/core";
import { useSelector } from "react-redux";

const useStyles = makeStyles((theme) => ({
  formControl: {
    width: "100%",
  },
}));

const getCurrentDate = () => {
  const now = new Date();
  return now.toISOString().slice(0, 10);
};

export default function DialogComponent({
  // open,
  handleDialogClose,
  handleSubmit,
  editTodo,
  isEditMode,
}) {
  const classes = useStyles();
  //These are local state hence there is no need to add it to redux.
  const [todoText, setTodoText] = useState("");
  const [priority, setPriority] = useState("Low");
  const [date, setDate] = useState(getCurrentDate());


  useEffect(() => {
    if (isEditMode) {

      setTodoText(editTodo.val);
      setPriority(editTodo.priority);
      setDate(editTodo.due);
    } else {

      setTodoText("");
      setPriority("Low");
      setDate(getCurrentDate());
    }
  }, [isEditMode]);

  const formSubmit = () => {
    let newTodo = {
      val: todoText,
      priority: priority,
      dueDate: date,
    };



    handleSubmit(newTodo);
    handleDialogClose();
    setTodoText("");
    setPriority("Low");
    setDate(getCurrentDate());
  };
  
  //Setting open tp false from the redux store 
  const open=useSelector((state) =>state.dialog.value);


  return (
    <Dialog open={open} onClose={handleDialogClose}>
      <DialogTitle>{isEditMode ? "Update" : "Add Todo"}</DialogTitle>
      <DialogContent>
        <Grid container direction="column" spacing={2}>
          <Grid item>
            <TextField
              label="Todo Text....."
              variant="outlined"
              name="todoText"
              onChange={(e) => setTodoText(e.target.value)}
              value={todoText}
              className={classes.formControl}
            />
          </Grid>
          <Grid item>
            <FormControl className={classes.formControl} variant="filled">
              <InputLabel id="priority">Priority</InputLabel>
              <Select
                name="priorityType"
                onChange={(e) => setPriority(e.target.value)}
                value={priority}
              >
                <MenuItem value="Low">Low</MenuItem>
                <MenuItem value="Med">Med</MenuItem>
                <MenuItem value="High">High</MenuItem>
              </Select>
            </FormControl>
          </Grid>
          <Grid item>
            <TextField
              id="date"
              label="Due Date"
              className={classes.formControl}
              type="date"
              value={date}
              onChange={(e) => setDate(e.target.value)}
              name="date"
              InputLabelProps={{
                shrink: true,
              }}
            />
          </Grid>
        </Grid>
      </DialogContent>
      <DialogActions>
        <Button onClick={handleDialogClose}>Cancel</Button>
        <Button color="primary" onClick={formSubmit}>
          {isEditMode ? "Update" : "Add"}
        </Button>
      </DialogActions>
    </Dialog>
  );
}
