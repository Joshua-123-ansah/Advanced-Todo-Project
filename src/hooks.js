import React from "react";

const getCurrentDate = () => {
  const now = new Date();
  return now.toISOString().slice(0, 10);
};

export const AllHooks = () => {
  const [todoText, setTodoText] = React.useState("");
  const [priority, setPriority] = React.useState("Low");
  const [date, setDate] = React.useState(getCurrentDate());

  return { todoText, setTodoText, priority, setPriority, date, setDate };
};
