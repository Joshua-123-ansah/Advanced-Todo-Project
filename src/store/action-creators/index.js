export const setOPenDialog = () => {
  return (dispatch) => (dispatch({ type: "setOPenDialog" }));
};

export const setCloseDialog = () => {
  return (dispatch) => (dispatch({ type: "setCloseDialog" }));
};

export const setIsEditMode = () => {
  return (dispatch) => (dispatch({ type: "setIsEditMode" }));
};

export const setCloseEditMode = () => {
  return (dispatch) => (dispatch({ type: "setCloseEditMode" }));
};

export const setMainTodos = (mainTodos) => {
  return (dispatch) => (dispatch({
    type: "setMainTodos",
    payload: mainTodos,
  }));
};

export const setTodos = (todos) => {
  return (dispatch) => (dispatch({
    type: "setTodos",
    payload: todos,
  }));
};
