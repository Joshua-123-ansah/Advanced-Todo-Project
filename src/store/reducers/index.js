import { combineReducers } from "redux";
import dialogReducer from "./Dialog";
import editReducer from "./Edit";
import mainTodoReducer from "./MainTodos";
import todosReducer from "./Todos";

const reducers = combineReducers({
  dialog: dialogReducer,
  edit: editReducer,
  mainTodo: mainTodoReducer,
  todos: todosReducer,
});

export default reducers;
