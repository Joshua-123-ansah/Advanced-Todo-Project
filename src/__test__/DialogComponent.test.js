import React from "react";
import { render as rtlRender, fireEvent } from "@testing-library/react";
import { renderHook, act } from "@testing-library/react-hooks";
import "@testing-library/jest-dom/extend-expect";
import { store } from "../store/store";
import { Provider } from "react-redux";
import DialogComponent from "../component/DialogComponent";
import { AllHooks } from "../hooks";

const render = (component) =>
  rtlRender(<Provider store={store}>{component}</Provider>);

describe("<DialogComponent />", () => {
  const editTodo = {};
  const currentDate = new Date().toISOString().slice(0, 10);

  test("render component", () => {
    const { container } = render(
      <DialogComponent
        handleDialogClose={() => {}}
        handleSubmit={() => {}}
        editTodo={editTodo}
      />
    );

    expect(container).toMatchSnapshot();
  });

  test("checking whether all states in the component has it initial values", () => {
    const { result } = renderHook(() => AllHooks());

    //todoText State
    expect(result.current.todoText).toEqual("");
    expect(result.current.priority).toEqual("Low");
    expect(result.current.date).toEqual(currentDate);
  });


  test("checking whether the setStates work", () => {
    const { result } = renderHook(() => AllHooks());

    //todoText State
    act(()=>{
      result.current.setTodoText("Read React Test")
      result.current.setPriority("High")
    })

    expect(result.current.todoText).toBe("Read React Test")
    expect(result.current.priority).toBe("High")
  });

  // test("test to ensure that the add or update button is clicked only once", () => {
  //   const { getByTestId } = render(
  //     <DialogComponent
  //       handleDialogClose={() => {}}
  //       handleSubmit={() => {}}
  //       editTodo={editTodo}
  //     />
  //   );

  //   const button=getByTestId("submit")
  // })
});
