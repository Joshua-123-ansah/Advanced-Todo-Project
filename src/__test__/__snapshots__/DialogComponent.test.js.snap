// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DialogComponent /> render component 1`] = `<div />`;
